capsules/extra/src/tickv.rs

```diff
@@ -270,37 +266,71 @@ impl<'a, F: Flash, H: Hasher<'a, 8>, const PAGE_SIZE: usize> flash::Client<F>
                 }
                 _ => {}
             },
-            Operation::GetKey => match ret {
-                Ok(tickv::success_codes::SuccessCode::Complete)
-                | Ok(tickv::success_codes::SuccessCode::Written) => {
-                    self.operation.set(Operation::None);
-                    self.client.map(|cb| {
-                        cb.get_value_complete(
-                            Ok(()),
-                            self.key_buffer.take().unwrap(),
-                            self.ret_buffer.take().unwrap(),
-                        );
-                    });
-                }
-                Err(tickv::error_codes::ErrorCode::EraseNotReady(_)) | Ok(_) => {}
-                _ => {
-                    self.operation.set(Operation::None);
-                    self.client.map(|cb| {
-                        cb.get_value_complete(
-                            Err(ErrorCode::FAIL),
-                            self.key_buffer.take().unwrap(),
-                            self.ret_buffer.take().unwrap(),
-                        );
-                    });
+            Operation::GetKey => {
+                buf_buffer.map(|buf| {
+                    self.ret_buffer.replace(buf);
+                });
+
+                match ret {
+                    Ok(tickv::success_codes::SuccessCode::Complete)
+                    | Ok(tickv::success_codes::SuccessCode::Written)
+                    | Err(tickv::error_codes::ErrorCode::BufferTooSmall(_)) => {
```

Comment:
Why do we signal this as a success to the caller?

Reply:
Because on delete our kv store layer only reads the header, not the entire value, so it only provides a smaller buffer.

Reply:
@alistair23

Reply:
I don't follow. What delete?

Reply:
`kv_store::delete()`.

To remove an object we call get with a buffer large enough to read just the tock tickv header.

Reply:
But this is the `Operation::GetKey` operation, does it call delete?

Reply:
https://github.com/tock/tock/blob/c91004925ef9699dc962940f4e67701db7de18c6/capsules/extra/src/kv_store.rs#L354

Can you look at the kv store implementation?

Reply:
Shouldn't the kv_store implementation handle this error then? It knows if a delete operation is happening

Reply:
I mean, the code does what the code does. This PR is to change it. Currently this error gets returned as FAIL. What is the caller supposed to do?

Why are we going back and forth? You can just try this and see the issue, no?

Reply:
I still don't follow.

So in `kv_store` the `generate_key_complete()` function calls `get_value()` with a hashed key and a buffer to store the value.

That ends up triggering the `read_complete()` callback for a `Operation::GetKey` in `tickv.rs`. 

The `ret` is an `Err(tickv::error_codes::ErrorCode::BufferTooSmall(_))` which triggers the `get_value_complete()` callback.

But then `get_value_complete()` doesn't check `result` because it expects a failure. So what doesn't work? The PR and commit messages don't really explain what this is fixing

Reply:
How can kv_store not check `result`? What if some other error occurs?

Reply:
Yeah, I agree it should check result.

A small change like this:

```diff
diff --git a/capsules/extra/src/tickv.rs b/capsules/extra/src/tickv.rs
index d70e4e13a..fa91f5319 100644
--- a/capsules/extra/src/tickv.rs
+++ b/capsules/extra/src/tickv.rs
@@ -282,6 +282,16 @@ impl<'a, F: Flash, H: Hasher<'a, 8>, const PAGE_SIZE: usize> flash::Client<F>
                         );
                     });
                 }
+                Err(tickv::error_codes::ErrorCode::BufferTooSmall(_)) => {
+                    self.operation.set(Operation::None);
+                    self.client.map(|cb| {
+                        cb.get_value_complete(
+                            Err(ErrorCode::SIZE),
+                            self.key_buffer.take().unwrap(),
+                            self.ret_buffer.take().unwrap(),
+                        );
+                    });
+                }
                 Err(tickv::error_codes::ErrorCode::EraseNotReady(_)) | Ok(_) => {}
                 _ => {
                     self.operation.set(Operation::None);
```

and then `get_value_complete()` can check the result and ignore the `ErrorCode::SIZE` as expected

Reply:
Can you create a branch or suggested change?

Reply:
Something like this? https://github.com/alistair23/tock/commit/76714971943bbd4d46e3a980748e34b3ddca94fa